xarray/core/datatree.py

```diff
@@ -926,7 +919,9 @@ def __setitem__(
         else:
             raise ValueError("Invalid format for key")
 
-    def update(self, other: Dataset | Mapping[str, DataTree | DataArray]) -> None:
+    def update(
+        self, other: Dataset | Mapping[str, DataTree | DataArray | Variable]
+    ) -> None:
         """
         Update this node's children and / or variables.
 
```

Comment:
This is a comment for L937 (can't add it directly)...

This is another `mypy` issue because the type of `k` can be a `Hashable` (from iterating through a `Dataset`) or a `str`, but `DataTree.name` needs a `str`.

There seem to be a couple of options:

1) Cast `k` as a string.
2) Check that `k` is a string, and raise an exception if it isn't.

Happy to do either (or something else). I couldn't think of an immediate issue with casting a `Hashable` to a string, but wanted to check (in case there might be some chance of a weird collision between e.g. `1` and `"1"`).

Reply:
This is symptomatic of a more general issue that technically all the keys in `DataTree` should be `Hashable` to match what xarray "mostly" supports (see https://github.com/pydata/xarray/pull/8294#issuecomment-1852714645).

@headtr1ck do you have any thoughts on supporting `Hashable` for names of child nodes in `DataTree`?

Reply:
I think we probably have to support `Hashable` at some point, otherwise any operation that combines a `DataTree` with a `Dataset` or `DataArray` will be a nightmare to type.

Reply:
Okay thanks. In that case I defer to Owen on whether it would be easier to do that in this PR or a follow-up one.

Reply:
I'm helping out while Owen is at a conference.  I was going to see if it were easy to add this change to handle Hashable and quickly ran into the `PurePosixPath`'s `pathsegments` needing to be `PathLike`.

I think if we want to use NodePath for traversing and altering the tree, the paths are going to need to be coerced into strings.

`DataTree`'s `_name` can be changed to `Hashable`, but when it's calling `NamedNode`'s `__init__` I think that property setter will have to coerce to a string.  But I'm not thinking of what kind of problems is that going to cause?
  Beyond say someone using `(7,)` as a key/path as well as `"(7,)"` and having collisions.  I think we can handle that by checking collisions. but I'm not 100% sure yet.  Am I missing something obvious?

This was going to be a question for tomorrow's meeting.

Reply:
In that meeting we discussed how having Hashables in datatree is possibly more trouble than it is worth, because (a) you can't form a valid path out of a list of hashables (but a list of strings is always fine), and (b) names of groups can't be hashable-like in netCDF or definitely not in Zarr anyway, so there doesn't seem like much of a use case for hashable-named groups at least.

For now it was decided to see if we could `type: ignore` our way to having an initial implementation that does not support hashables in datatree (which as we can explicitly forbid hashables at tree creation time hopefully isn't a ridiculous idea). #8836 was made to track the intent to revisit this.

Reply:
I did not really follow all the discussions about this PR.
But for me, not accepting Hashables sounds reasonable (even though just the thought of a tree-like datastructure sounds like this should be possible for any Hashables names. E.g. a node object with a Hashables name and a number of node children). But agreed, it makes things like getitem unnecessarily complicated.

Probably the best for now is some type ignores or casts.
In the future I anyway have plans to make Dataset a generic class in variable names (and dimension names for that matter). Then this problem can be solved by returning, e.g. `Dataset[str]`.

Reply:
> I did not really follow all the discussions about this PR.

Yeah sorry - a lot of them happened over zoom (see the March 19th [meeting notes here](https://hackmd.io/tyqxf4dkTFaRw9Udq1PYog)).

> sounds like this should be possible for any Hashables names

The problem is not the tree construction, it's serialization (because I can't guarantee being able to make a single valid unix path out of .join'ing all those hashables).

> Probably the best for now is some type ignores or casts.

Cool, that's what we've done.

> In the future I anyway have plans to make Dataset a generic class in variable names (and dimension names for that matter). Then this problem can be solved by returning, e.g. `Dataset[str]`.

That would be awesome!